### DISCUSSION TITLE
Commit rules - to have or not to have

### DISCUSSION THREAD

**Discussion by leamas** [opencpn-manuals/main#3] [General]
May I suggest some simple rules for commits in the main branch? Since we now are three users, we might need to discuss these things so we can find a common understanding. Some suggestions:

  - At a least a complete sentence describing each commit?
  - That we use the standard pre-commit hook to avoid space problems?
  - That we avoid "test" commits on the main branch and uses temporary branches for the testing we obviously should do?
  - EDIT: That we use recommended pre-commit hooks.

Thoughts?

**Comment by Rasbats** [comment]
>use the standard pre-commit hook

A reminder of what is needed please! Can github desktop be used with this?

**Reply by leamas** [reply]
github desktop make no difference. But activate on the command line:

    > cd .git/hooks
    > copy pre-commit.sample pre-commit

And that's that... the pre-commit hook will check your next commit (and most likely find errors...). You can commit despite errors using git commit --no-verify, but generally it's not  a good idea...

**Comment by leamas** [comment]
> That we avoid "test" commits on the main branch and uses temporary branches for the testing we obviously should do?

Note that with new setup there is no need to push test commits. Something like this should work:

    > git checkout -b foo    # checkout a new branch foo
    > make-all-sorts-of-test-commits
    > git push -f origin  foo:gh-pages

This will make a complete build without committing anything on the  main branch. When things are in order, one can make tested commits on main branch and push it.

**Comment by rgleason** [comment]
Thank you! 
Does a precommit hook need to be refreshed or "updated"?  OR does the hook look online to github.com to some file that is always current, when it is callled into action?

**Reply by leamas** [reply]
It needs to be refreshed, so to speak. The copy in `.git/hooks/pre-commit ` is _not_ part of the repo.  What we have is the file `pre-commit` in the root directory. This is updated when git updates the repo, but is not used. To use the latest version one need to something like `cp pre-commit .git/hooks/pre-commit`.

This is for the plugins repo. Other repos uses the standard git pre-commit, which is named pre-commit.sample. For now, this is what we use in the manual repos. As long as we just use this, I hardly see a need to update the hook.

Bottom line: A git hook is a personal thing for the developer, not a part of the git repo. The git repo might or might not have a suggested variant of the hook, but user must copy it to .git/hooks for it to be actually used.

**Comment by rgleason** [comment]
Thanks. This is very helpful. Can it be included in the commit rules?

**Reply by leamas** [reply]
I have updated that to include the fact that we should actually use the pre-commit hook; that is the rule part which was missing.

What we have here is the "How" part. I'm currently writing on a git workflows document, it might go into it.

**Comment by rgleason** [comment]
https://github.com/opencpn-manuals/development/commit/3f4a331911f40b92c7eebda83f6a8ef19ebf2685
